keymanapp/keyman
# refactor(web): mouse-based use of touch handlers 👆

Continuing from #5506, this PR implements a mouse-based version of the new abstraction, `InputEventEngine`.  Adding that to the other changes from #5506, this allows us to drop the old mouse event handlers and replace them with the handlers that were originally touch-oriented!

This is most easily seen by clicking down on an OSK key (in non-touch / 'desktop' mode) then dragging the pointer around the OSK without lifting; the key underneath the mouse will automatically highlight, just like our standard touch keyboard behavior.  No key previews here - it's a 'desktop' layout, not a 'phone' layout - previews are 'phone' only.  Since the desktop keyboard layout also doesn't do subkeys, there's not much else to directly show for it at the moment without hacking KMW a little bit.

Of course, this does get us _much_ closer to having an interactive OSK independent of the actual device that KMW has been loaded on.  It's now simply a matter of actually _supporting_ loading & display of a touch OSK while on a 'desktop' device.

------

This also does preserve the pre-existing key-gap behavior for mice:  clicking the gap between two keys on the same row will not auto-select one.  That said... it's different when clicking the gap between _rows_, as _that_ gap has always been considered part of the key itself internally.

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/25213402/127622189-b14ec674-a845-492a-84da-71d06087508d.png">

The part in blue is considered part of the key square, and the key-lookup operates by looking for the key-square, rather than the key it contains.  Remedying this would take additional work that's likely best left for another time.  Orange is the key's margin, space _outside_ of the key's element but related to it.

## User Testing

@keymanapp/testers 

### Suite 1:  `desktop` form factor tests

**Platform / browser combinations**:
- [ ] Windows / Edge
- [ ] macOS / Safari
    - Unless otherwise specified, use the "**Test unminified KeymanWeb**" test page.

<details>
  <summary>Tests (with inlined instructions)</summary>

- [ ] TEST.RESIZING:  Using the resize handle (<img width="28" alt="image" src="https://user-images.githubusercontent.com/25213402/126436281-2ea2c834-4fb0-45b4-b57a-8e3a9dcd817f.png">) for the desktop OSK, resize it at least three different ways.
     - Each time, use the OSK to type 'a', 'p', and then 'e'.  Confirm the correct character is output.
- [ ] TEST.HELD_MOUSE: In sequence...
    1. Click and hold the OSK's 'p' key.
    2. Without releasing the left mouse button, move the cursor across the top row to 'q', then down to 'a'.
        - Confirm that the key underneath the cursor at any point in time is the only one highlighted.
    3. Release the 'a' key.
        - Confirm that an 'a' is output.

</details>
